File: 2021/python/609.find-duplicate-file-in-system.py
class Solution:
    def findDuplicate(self, paths: List[str]) -> List[List[str]]:
        if not paths or len(paths) == 0:
            return list()
        
        res = list()
        file_dict = collections.defaultdict(list)
        
        for path in paths:
            p_dir, *files = path.split(' ')
            for file in files:
                file_name, content = file.split('(')
                file_dict[content].append(p_dir + '/' + file_name)
        
        for _, file_paths in file_dict.items():
            if len(file_paths) > 1: # add more than 2 files which has same content into res
                res.append(file_paths)
                
        return res

File: 2021/609.find-duplicate-file-in-system.go
func findDuplicate(paths []string) [][]string {
    fileDict := make(map[string][]string)
    
    for _, path := range paths {
        idx := strings.IndexByte(path, ' ')
        
        p_dir, files := path[:idx], path[idx + 1:]
        for _, file := range strings.Split(files, " ") {
            fIdx := strings.IndexByte(file, '(')
            fileName, content := file[:fIdx], file[fIdx + 1:]
            fileDict[content] = append(fileDict[content], p_dir + "/" + fileName)
        }
    }
    
    res := [][]string{} 
    for _, group := range fileDict {
        if len(group) > 1 {
            res = append(res, group)
        }   
    }
    return res
}
